chai_lab/data/dataset/msas/preprocess.py

```diff
@@ -55,168 +21,82 @@ def merge_main_msas_by_chain(msas: list[MSAContext]) -> MSAContext:
-        if len(msa_id2sequence) < num_msas_to_pair:
-            continue  # not present in one of non-empty MSAs
-
-        n_pairings = min(len(seqs) for seqs in msa_id2sequence.values())
-
-        # select closest by edit distance
-        chunk = torch.zeros([len(msas), n_pairings], dtype=torch.long)
-        mask = torch.zeros([len(msas), n_pairings], dtype=torch.bool)
-
-        for msa_id, msa in enumerate(msas):
-            if msa_id in msa_id2sequence:
-                seq_idx = torch.asarray(msa_id2sequence[msa_id], dtype=torch.long)
-                edit_dists = msa_similarity[msa_id][seq_idx]
-
-                chunk[msa_id, :] = seq_idx[torch.argsort(edit_dists)][:n_pairings]
-                mask[msa_id, :] = True
-
-        chunks.append(chunk)
-        masks.append(mask)
-
-    # num_msas x sum_num_indices
-    return torch.concatenate(chunks, dim=1), torch.concatenate(masks, dim=1)
 
+def drop_duplicates(msa: MSAContext) -> MSAContext:
+    # when chains are cropped out, they have zero tokens. We should not try to
+    # deduplicate in this case.
+    if msa.tokens.shape[-1] == 0:
+        return msa
+    if msa.depth == 0:
+        return msa
+    # sorted: don't shuffle the paired and main MSA
+    _, idxs = unique_indexes(msa.tokens, dim=0, sorted=True)
+    return msa[idxs, :]
 
-def pair_msas_by_chain_with_species_matching(msas: list[MSAContext]) -> MSAContext:
-    """
-    Combines the input MSAs (from different chains) across the different chains into a
-    single MSA, merging chains along the tokens dimension and pairing by pairing key.
-    """
-    if len(msas) == 0:
-        logging.warning("No MSAs to pair; they are all empty")
-        return MSAContext.create_empty(0)
-    assert all(msa.num_tokens > 0 for msa in msas)
 
-    # Create paired part of the MSA
-    all_stacked_matches, all_stacked_masks = [
-        x[:, :MAX_PAIRED_DEPTH]
-        for x in _pair_msas_by_chain_with_species_matching_just_pairing(msas)
-    ]
-    num_msas, sum_num_indices = all_stacked_matches.shape
-    assert num_msas == len(msas)
-
-    # Truncate the MSAs to the query sequence
-    query_msas = [msa[0:1, :] for msa in msas]
-
-    # Get the paired sequences for each MSA (if we found any species matches)
-    paired_msas: list[MSAContext] = []
-    for msa_idx, (matches, mask) in enumerate(
-        zip(all_stacked_matches, all_stacked_masks, strict=True)
-    ):
-        msa_chunk = msas[msa_idx][matches, :]
-        mask = rearrange(mask, "depth -> depth 1")
-        paired_msas.append(msa_chunk.apply_mask(mask))
-
-    # Stack the query and paired MSAs for each chain along the depth dimension
-    full_msas = [
-        MSAContext.cat(
-            [query_msa, paired_msa],
-            dataset_source=MSADataSource.PAIRED,
-            dim=0,
-        )
-        for query_msa, paired_msa in zip(query_msas, paired_msas)
+def pair_and_merge_msas(msas: list[MSAContext]) -> MSAContext:
```

Comment:
When pairing, we want to first sort by sequence similarity to query before matching up within the pairing key. This _should_ implicitly happen because `a3m` files are sorted by similarity and `aligned.pqt` is built from `a3m`, but it's probably better to handle this explicitly; I don't see logic for this now.

Reply:
I assume that's in a3m creation logic; fine to add this as an explicit step